### PR TITLE
[RFR] New ra-tree for react-admin v2

### PR DESCRIPTION
**Pull request by djhi**
- [x] Refactor according to feedbacks
- [x] Actions on nodes
- [x] Add drag & drop with reordering
- [x] Optimistic updates
- [x] Optimistic updates for react-admin actions
- [x] Adding new nodes
- [x] Tests
- [x] Documentation

ra-tree-drag-drop

### CODE REVIEW

File: packages/ra-tree-core/README.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ import { TreeController } from 'ra-tree-core';
 import TreeNode from './TreeNode';
 
 export const CategoriesList = (props) => (
-    <TreeController {...props}>
+    <TreeController resource="categories" {...props}>
```
Comment (djhi) [inline packages/ra-tree-core/README.md]: You don't need this prop, it's passed on by react-admin like with a classic `List`

File: packages/ra-tree-core/src/reducers/data.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
+import { Reducer } from 'redux';
```
Comment (fzaninotto) [inline packages/ra-tree-core/src/reducers/data.ts]: You're duplicating most of the logic of the ra-core data reducer. Why can't you just store the extra data that you need in this reducer?

File: packages/ra-tree-core/src/reducers/nodes.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
+import { DELETE, DELETE_MANY, Identifier } from 'ra-core';
```
Comment (fzaninotto) [inline packages/ra-tree-core/src/reducers/nodes.ts]: The code of this file is very hard to read and understand. Maybe by introducing temporary constants, that would make the code clearer? Or maybe extract some of the processing into functions.

File: packages/ra-tree-core/src/selectors.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { ROOT_NODE_ID } from './constants';
+
+export const getTreeRootNodes = (state, resource) =>
```
Comment (fzaninotto) [inline packages/ra-tree-core/src/selectors.ts]: a jsdoc with an example state would reaaaaaally help understanding the reducers

File: packages/ra-tree-language-french/index.js [hunk 5]
```diff
@@ -1,9 +1,11 @@
 module.exports = {
     ra: {
         tree: {
-            drag_preview:
-                'Nœud #%{id} |||| Nœud #%{id} contenant %{smart_count} enfants',
-            root_target: 'Lâcher un nœud ici pour le remonter à la racine',
+            add_child_node: 'Ajouter un enfant',
+            add_node_after: 'Insérer un noeud après',
```
Comment (fzaninotto) [inline packages/ra-tree-language-french/index.js]:
```suggestion
            add_node_after: 'Insérer un nœud après',
```

File: packages/ra-tree-language-french/index.js [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
+            add_node_before: 'Insérer un noeud avant',
```
Comment (fzaninotto) [inline packages/ra-tree-language-french/index.js]:
```suggestion
            add_node_before: 'Insérer un nœud avant',
```

File: packages/ra-tree-ui-materialui/README.md [hunk 7]
```diff
@@ -67,174 +68,85 @@ const App = () => (
...
-    - `record`: the node's original data
-  - any additional props received by the `TreeController` component
 - `parentSource`: The field used as the parent identifier for each node. Defaults to `parent_id`
+- `positionSource`: The field used to sort the nodes. Optional.
 
-### <NodeView>
+### <TreeList>
 
-The `NodeView` component accepts the following props:
+This is the UI container. Does nothing for now but may includes advanced selection features in the future. Takes no props.
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/README.md]:
```suggestion
This is the UI container. Does nothing for now but may include advanced selection features in the future. Takes no props.
```

File: packages/ra-tree-ui-materialui/src/AddNodeAfterMenuItem.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    parentSource,
+    positionSource,
+    record: siblingRecord,
+    translate,
+    ...props
+}) => {
+    const record = {
+        [parentSource]: siblingRecord[parentSource],
+        [positionSource]: siblingRecord[positionSource] + 1,
+    };
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/AddNodeAfterMenuItem.tsx]: I don't understand that. Are you partially cloning a record? For what? An explanation and an example in jsDoc would help.

File: packages/ra-tree-ui-materialui/src/AddNodeBeforeMenuItem.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    parentSource,
+    positionSource,
+    record: siblingRecord,
+    translate,
+    ...props
+}) => {
+    const record = {
+        [parentSource]: siblingRecord[parentSource],
+        [positionSource]: siblingRecord[positionSource],
+    };
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/AddNodeBeforeMenuItem.tsx]: Same

File: packages/ra-tree-ui-materialui/src/DeleteWithConfirmMenuItem.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
+import React, { Fragment, Component, ReactElement } from 'react';
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/DeleteWithConfirmMenuItem.tsx]: This should be in ra-ui-material-ui. Same with other menu items.

File: packages/ra-tree-ui-materialui/src/Tree.tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        ...props
+    }: Props & InjectedProps & ViewProps & WithStyles<typeof styles>) => (
+        <div
+            className={classnames('tree-page', classes.root, className)}
+            {...sanitizeRestProps(props)}
+        >
+            <Title title={title} defaultTitle={defaultTitle} />
+            <Card className={classes.card}>
+                {actions && (
+                    <TreeListToolbar
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.tsx]: how is that different from a `ListToolbar`?

File: packages/ra-tree-ui-materialui/src/Tree.tsx [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    )
+);
+
+const warnAboutChildren = () =>
+    console.warn(
+        // eslint-disable-line
+        `You passed multiple children to the Tree component.
+
+    <Tree>
+        <NodeView>
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.tsx]: does not exist anymore

File: packages/ra-tree-ui-materialui/src/Tree.tsx [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+            <TextField source="name" />
+        </NodeView>
+    </Tree>
+
+    // Or
+
+    <Tree>
+        <NodeForm>
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.tsx]: does not exist anymore

File: packages/ra-tree-ui-materialui/src/Tree.tsx [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+            <TextInput source="name" />
+        </NodeForm>
+    </Tree>
+
+If you need actions on each node, use the actions prop on either the NodeView or NodeForm component:
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.tsx]: I think this must be rewritten

File: packages/ra-tree-ui-materialui/src/TreeListToolbar.tsx [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+const TreeListToolbar = ({
+    classes,
+    filters,
+    permanentFilter, // set in the List component by the developer
+    actions,
+    exporter,
+    ...rest
+}) => (
+    <Toolbar className={classes.toolbar}>
+        <span />
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeListToolbar.tsx]: I understand you're using ` justifyContent: 'space-between'` and this `span` to push the actions to the right. But that means that 90% of the line is the toolbar... Wrong solution IMHO.

File: packages/ra-tree-ui-materialui/src/TreeNode.tsx [hunk 16]
```diff
@@ -0,0 +1,538 @@
+import React, {
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeNode.tsx]: this file needs to be split into several smaller components, it's too hard to follow

File: packages/ra-tree-core/src/reducers/data.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-tree-core/src/reducers/data.ts]: I would have to dispatch a *fake* `CRUD_LIST_SUCCESS` for that but that could work

File: packages/ra-tree-ui-materialui/src/Tree.tsx (hunk 11 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/Tree.tsx]: Filters

File: packages/ra-tree-ui-materialui/src/TreeNodeActions.tsx [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        root: {
+            marginLeft: 'auto',
+            paddingRight: theme.spacing.unit * 3,
+        },
+    });
+
+const TreeNodeActions = withStyles(styles)(
+    ({
+        basePath,
+        children,
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeNodeActions.tsx]: Documented in this package readme at L139

File: packages/ra-tree-ui-materialui/src/DeleteWithConfirmMenuItem.tsx (hunk 10 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/DeleteWithConfirmMenuItem.tsx]: Do we need them elsewhere ?

File: packages/ra-tree-ui-materialui/src/TreeListToolbar.tsx (hunk 15 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeListToolbar.tsx]: That's how it is done in the ListToolbar from which I just removed the filters
